File: parsers/kotlin/src/main/kotlin/verity/passport/prover/PassportError.kt
package verity.passport.prover

/**
 * Sealed hierarchy of errors that can occur while parsing or verifying an ePassport document.
 *
 * All errors extend [IllegalArgumentException] following the convention established by
 * kotlinx.serialization's `SerializationException`: these failures are fundamentally caused
 * by invalid or unexpected input data, not by programming mistakes.
 *
 * Domain-specific data is exposed as typed `val` properties so callers can inspect the
 * failing value without parsing the message string.
 */
public sealed class PassportError(message: String) : IllegalArgumentException(message) {

    /** Indicates that the hash of DG1 stored in eContent does not match the computed hash. */
    public class Dg1HashMismatch : PassportError("DG1 hash mismatch in eContent")

    /** Indicates that the hash of eContent stored in SignedAttributes does not match the computed hash. */
    public class EcontentHashMismatch : PassportError("eContent hash mismatch in SignedAttributes")

    /** Indicates that a data item exceeded the maximum allowed buffer size. */
    public class BufferOverflow(
        public val detail: String,
    ) : PassportError("Data too large for buffer: $detail")

    /** Indicates that a required piece of data could not be located in the structure. */
    public class DataNotFound(
        public val detail: String,
    ) : PassportError("Required data not found: $detail")

    /** Indicates that the signature algorithm OID is not supported by this implementation. */
    public class UnsupportedSignatureAlgorithm(
        public val algorithm: String,
    ) : PassportError("Unsupported signature algorithm: $algorithm")

    /** Indicates a failure while decoding the CMS (Cryptographic Message Syntax) structure. */
    public class CmsParsingFailed(
        public val detail: String,
    ) : PassportError("CMS parsing failed: $detail")

    /** Indicates a failure while decoding an ASN.1 structure. */
    public class Asn1DecodingFailed(
        public val detail: String,
    ) : PassportError("ASN.1 decoding failed: $detail")

    /** Indicates that a mandatory field was absent from the parsed structure. */
    public class MissingRequiredField(
        public val field: String,
    ) : PassportError("Missing required field: $field")

    /** Indicates that the DG1 hash entry is absent from the eContent hash table. */
    public class MissingDg1Hash : PassportError("Missing DG1 hash in eContent")
}

File: parsers/swift/Sources/PassportProverLib/Models/PassportError.swift
import Foundation

public enum PassportError: LocalizedError, Sendable, Equatable {
    case dg1HashMismatch
    case econtentHashMismatch
    case invalidDscKey
    case dscSignatureInvalid
    case cscaSignatureInvalid
    case bufferOverflow(String)
    case rsaExponentTooLarge
    case dataNotFound(String)
    case unsupportedSignatureAlgorithm(String)
    case cmsParsingFailed(String)
    case x509ParsingFailed(String)
    case asn1DecodingFailed(String)
    case missingRequiredField(String)
    case missingDg1Hash

    public var errorDescription: String? {
        switch self {
        case .dg1HashMismatch: return "DG1 hash mismatch in eContent"
        case .econtentHashMismatch: return "eContent hash mismatch in SignedAttributes"
        case .invalidDscKey: return "Invalid DSC public key"
        case .dscSignatureInvalid: return "DSC signature verification failed"
        case .cscaSignatureInvalid: return "CSCA signature verification failed"
        case .bufferOverflow(let d): return "Data too large for buffer: \(d)"
        case .rsaExponentTooLarge: return "RSA exponent too large"
        case .dataNotFound(let d): return "Required data not found: \(d)"
        case .unsupportedSignatureAlgorithm(let a): return "Unsupported signature algorithm: \(a)"
        case .cmsParsingFailed(let d): return "CMS parsing failed: \(d)"
        case .x509ParsingFailed(let d): return "X.509 certificate parsing failed: \(d)"
        case .asn1DecodingFailed(let d): return "ASN.1 decoding failed: \(d)"
        case .missingRequiredField(let f): return "Missing required field: \(f)"
        case .missingDg1Hash: return "Missing DG1 hash in eContent"
        }
    }
}
